### PR TITLE
feat(npu_compiler): npu compiler opt

### PR DESCRIPTION
**Pull request by mingfa-byted**
feat(CI): add npu github workflow
	feat(pre-commit-config): add python code format
	feat(npu-opt): fuse_brc_permute_sum
	feat(npu-opt): fused_div_mul_sum

### CODE REVIEW

File: xpu_graph/passes/patterns/targets/npu/__init__.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        for name in dir(module):
+            pat = getattr(module, name)
+            if (
+                isinstance(pat, type)
+                and issubclass(pat, Pattern)
+                and pat not in (Pattern, AutoMatchPattern)
+                and pat._opt_level <= config.opt_level
+            ):
+                for stage in pat._stages:
+                    patterns[pat._pattern_group].append(pat(stage))
```
Comment (zhangjihang-BD) [inline xpu_graph/passes/patterns/targets/npu/__init__.py:28]: 这个最新的提交 https://github.com/BD-Seed-HHW/xpu_graph/commit/bcb149dce78a7ea2c1be2cfc153ba363f98aeb46
改过了，要稍微调整下

File: xpu_graph/passes/patterns/targets/npu/triton_kernel/fused_div_mul_sum.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
+import torch
+import torch_npu
+import triton
+import triton.language as tl
+
+@triton.jit
+def fused_div_mul_sum_kernel(
```
Comment (zhangjihang-BD) [inline xpu_graph/passes/patterns/targets/npu/triton_kernel/fused_div_mul_sum.py:7]: 这个kernel不太好读了，要么改下变量名，要么加点comments吧

File: xpu_graph/passes/patterns/targets/npu/fuse_div_mul_sum.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            div2_base_input = expand2_node.args[0]
+            expand2_shape = expand2_node.args[1]
+
+            with graph.inserting_before(final_mul):
+                fused_node = graph.call_module(
+                    "npu_triton_fused_div_mul_sum",
+                    args=(
+                        div1_dividend,   # unsqueeze_2 的输出
+                        div1_base_input, # clamp_min 原始输入
+                        # expand1_shape,   # [10,15,1,4,128]
```
Comment (zhangjihang-BD) [inline xpu_graph/passes/patterns/targets/npu/fuse_div_mul_sum.py:126]: 删一下

File: xpu_graph/passes/patterns/targets/npu/fuse_div_mul_sum.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            final_mul, sum_node, mul_node, 
+            div1, expand1_node,
+            div2, expand2_node
+        ]
+
+
+    def process(self, gm: fx.GraphModule):
+        graph = gm.graph
+        changed = False
+        gm.add_submodule("npu_triton_fused_div_mul_sum", DivMulSumOperation())
```
Comment (zhangjihang-BD) [inline xpu_graph/passes/patterns/targets/npu/fuse_div_mul_sum.py:100]: 每次process都会加？最好做个has_submodule的判断

File: xpu_graph/cache.py [hunk 5]
```diff
@@ -1,17 +1,26 @@
...
-from .utils import logger
-from .fx_utils import FxStage
+
 import torch
 from torch._dynamo.convert_frame import compile_lock
 from torch.utils._python_dispatch import _disable_current_modes
-from torch._inductor.codecache import CompiledFxGraph, PyCodeCache, get_path
+
+torch_version = torch.__version__
+if torch_version.startswith("2.6"):
```
Comment (yfchen-byted) [inline xpu_graph/cache.py:12]: 注意下方CompiledFxGraph反序列化的地方，在pytorch2.5时需要额外post_compile以使能cudagraphs，在2.6时需要确认反序列化时是否就已经使能了cudagraphs，以及是否需要再调用postcompile

File: xpu_graph/passes/patterns/targets/npu/__init__.py (hunk 1 above)
Comment (wxue1) [reply xpu_graph/passes/patterns/targets/npu/__init__.py]: done

File: xpu_graph/passes/patterns/targets/npu/triton_kernel/fused_div_mul_sum.py (hunk 2 above)
Comment (mingfa-byted) [reply xpu_graph/passes/patterns/targets/npu/triton_kernel/fused_div_mul_sum.py:7]: 已经添加注释

**Review by zhangjihang-BD — APPROVED** [review]
